Firestore/core/src/firebase/firestore/local/leveldb_transaction.h

```diff
@@ -175,6 +177,14 @@ class LevelDbTransaction {
    */
   void Put(std::string key, std::string value);
 
+  /**
+   * Schedules the row identified by `key` to be set to `value` when this
+   * transaction commits.
+   */
+  void Put(std::string key, const nanopb::ByteString& value) {
+    Put(std::move(key), nanopb::MakeString(value.get()));
+  }
+
```

Comment:
> Excellent point on avoiding conversion, thanks!
> 
> I have added a `GrpcByteBufferWriter` class, though in the current state of the code, it doesn't really help with the tight coupling problem:
> 
> * `GrpcByteBufferWriter` is defined in `writer.{h,cc}` files, similar to other writers;
> * I have made `ToByteBuffer`, etc. free functions rather than member functions on `Message`, but they are still defined in the same file (I think helpers like these _should_ exist somewhere; dealing with writers and readers directly is annoying).

For Writers especially it's much more common to do a one shot conversion. I agree that it's reasonable to make functions for this.

> What are your thoughts re. where this translation point between gRPC and Nanopb should end up?

I think this should end up alongside our other gRPC machinery (i.e. as a peer of `grpc_completion.{h,cc}`.

> Incidentally, `Write/ReadNanopbMessage` function names seem like they need an update: `Message` is now confusing due to the newly-added class, and `Nanopb` seems redundant. Maybe just `Write/ReadProto`?

SG. These could even just be `Read` and `Write`.